core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -61,21 +76,68 @@ public void close() {
-     * @param timeUnit the time unit of the {@code timeout} argument
+     * @param timeout   the maximum time to wait for the in-flight operations block
+     * @param timeUnit  the time unit of the {@code timeout} argument
      * @param onBlocked the action to run once the block has been acquired
-     * @throws InterruptedException if calling thread is interrupted
-     * @throws TimeoutException if timed out waiting for in-flight operations to finish
+     * @param <E>       the type of checked exception thrown by {@code onBlocked}
+     * @throws InterruptedException      if calling thread is interrupted
+     * @throws TimeoutException          if timed out waiting for in-flight operations to finish
      * @throws IndexShardClosedException if operation permit has been closed
      */
-    public <E extends Exception> void blockOperations(long timeout, TimeUnit timeUnit, CheckedRunnable<E> onBlocked) throws
-        InterruptedException, TimeoutException, E {
+    <E extends Exception> void syncBlockOperations(
+            final long timeout,
+            final TimeUnit timeUnit,
+            final CheckedRunnable<E> onBlocked) throws InterruptedException, TimeoutException, E {
         if (closed) {
             throw new IndexShardClosedException(shardId);
         }
+        delayOperations();
+        doBlockOperations(timeout, timeUnit, onBlocked);
+    }
+
+    /**
+     * Immediately delays operations and on another thread waits for in-flight operations to finish and then executes {@code onBlocked}
+     * under the guarantee that no new operations are started. Delayed operations are run after {@code onBlocked} has executed. After
+     * operations are delayed and the blocking is forked to another thread, returns to the caller.
+     *
+     * @param timeout   the maximum time to wait for the in-flight operations block
+     * @param timeUnit  the time unit of the {@code timeout} argument
+     * @param onBlocked the action to run once the block has been acquired
+     * @param <E>       the type of checked exception thrown by {@code onBlocked} (not thrown on the calling thread)
+     */
+    <E extends Exception> void asyncBlockOperations(final long timeout, final TimeUnit timeUnit, final CheckedRunnable<E> onBlocked) {
+        delayOperations();
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+            try {
+                doBlockOperations(timeout, timeUnit, onBlocked);
+            } catch (final Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    private void delayOperations() {
+        synchronized (this) {
+            if (delayed) {
+                throw new IllegalStateException("operations are already delayed");
+            } else {
+                delayed = true;
+            }
+        }
+    }
+
+    private <E extends Exception> void doBlockOperations(
```

Comment:
better yet, how about the following? This way it's clear how a `delayOperations` call is matched by the undelay logic.

```
diff --git a/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java b/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
index 7fc56a1..f1c8b0d 100644
--- a/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
+++ b/core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java
@@ -27,6 +27,7 @@ import org.elasticsearch.action.support.ThreadedActionListener;
 import org.elasticsearch.common.CheckedRunnable;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.common.util.concurrent.ThreadContext.StoredContext;
 import org.elasticsearch.threadpool.ThreadPool;
 
@@ -95,7 +96,11 @@ final class IndexShardOperationPermits implements Closeable {
             throw new IndexShardClosedException(shardId);
         }
         delayOperations();
-        doBlockOperations(timeout, timeUnit, onBlocked);
+        try {
+            doBlockOperations(timeout, timeUnit, onBlocked);
+        } finally {
+            releasedDelayedOperations();
+        }
     }
 
     /**
@@ -110,12 +115,21 @@ final class IndexShardOperationPermits implements Closeable {
      */
     <E extends Exception> void asyncBlockOperations(final long timeout, final TimeUnit timeUnit, final CheckedRunnable<E> onBlocked) {
         delayOperations();
-        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
-            try {
-                doBlockOperations(timeout, timeUnit, onBlocked);
-            } catch (final Exception e) {
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(new AbstractRunnable() {
+            @Override
+            public void onFailure(Exception e) {
                 throw new RuntimeException(e);
             }
+
+            @Override
+            protected void doRun() throws Exception {
+                doBlockOperations(timeout, timeUnit, onBlocked);
+            }
+
+            @Override
+            public void onAfter() {
+                releasedDelayedOperations();
+            }
         });
     }
 
@@ -150,25 +164,30 @@ final class IndexShardOperationPermits implements Closeable {
                 throw new TimeoutException("timed out during blockOperations");
             }
         } finally {
-            final List<ActionListener<Releasable>> queuedActions;
-            synchronized (this) {
-                queuedActions = delayedOperations;
-                delayedOperations = null;
-                delayed = false;
-            }
-            if (queuedActions != null) {
-                // Try acquiring permits on fresh thread (for two reasons):
-                // - blockOperations can be called on recovery thread which can be expected to be interrupted when recovery is cancelled.
-                //   Interruptions are bad here as permit acquisition will throw an InterruptedException which will be swallowed by
-                //   ThreadedActionListener if the queue of the thread pool on which it submits is full.
-                // - if permit is acquired and queue of the thread pool which the ThreadedActionListener uses is full, the onFailure
-                //   handler is executed on the calling thread. This should not be the recovery thread as it would delay the recovery.
-                threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
-                    for (ActionListener<Releasable> queuedAction : queuedActions) {
-                        acquire(queuedAction, null, false);
-                    }
-                });
-            }
+            releasedDelayedOperations();
+        }
+    }
+
+    private void releasedDelayedOperations() {
+        final List<ActionListener<Releasable>> queuedActions;
+        synchronized (this) {
+            assert delayed;
+            queuedActions = delayedOperations;
+            delayedOperations = null;
+            delayed = false;
+        }
+        if (queuedActions != null) {
+            // Try acquiring permits on fresh thread (for two reasons):
+            // - blockOperations can be called on recovery thread which can be expected to be interrupted when recovery is cancelled.
+            //   Interruptions are bad here as permit acquisition will throw an InterruptedException which will be swallowed by
+            //   ThreadedActionListener if the queue of the thread pool on which it submits is full.
+            // - if permit is acquired and queue of the thread pool which the ThreadedActionListener uses is full, the onFailure
+            //   handler is executed on the calling thread. This should not be the recovery thread as it would delay the recovery.
+            threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+                for (ActionListener<Releasable> queuedAction : queuedActions) {
+                    acquire(queuedAction, null, false);
+                }
+            });
         }
     }
```